freezer/freezer.go

```diff
@@ -0,0 +1,178 @@
+package freezer
+
+import (
+	"context"
+	"time"
+
+	"github.com/uw-labs/freezer"
+	"github.com/uw-labs/straw"
+	"github.com/uw-labs/substrate"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	_ substrate.AsyncMessageSink   = (*asyncMessageSink)(nil)
+	_ substrate.AsyncMessageSource = (*asyncMessageSource)(nil)
+)
+
+type AsyncMessageSinkConfig struct {
+	streamstore straw.StreamStore
+	fconfig     freezer.MessageSinkConfig
+}
+
+func NewAsyncMessageSink(config AsyncMessageSinkConfig) (substrate.AsyncMessageSink, error) {
+	fms, err := freezer.NewMessageSink(config.streamstore, config.fconfig)
+	if err != nil {
+		return nil, err
+	}
+	return &asyncMessageSink{fms}, nil
+}
+
+type asyncMessageSink struct {
+	fms *freezer.MessageSink
+}
+
+func (ams *asyncMessageSink) PublishMessages(ctx context.Context, acks chan<- substrate.Message, messages <-chan substrate.Message) (rerr error) {
+	var toAck []substrate.Message
+	t := time.NewTimer(0)
+	t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case m := <-messages:
+			t.Reset(1000 * time.Millisecond)
```

Comment:
This timer business seems a bit funny to me. This call to `Reset` might be in a race with a write to the channel. They say in the docs that it should be avoided, but I'm not sure what are the consequences of that https://golang.org/pkg/time/#Timer.Reset 

I implemented this resetable ticker that should call `Reset` properly, so we can maybe pull it out if it would be useful here https://github.com/utilitywarehouse/partner-pkg/blob/master/ticker/ticker.go

Reply:
Fixed, I think.